src/caffe/layers/cudnn_conv_layer.cpp

```diff
@@ -111,48 +152,73 @@ void CuDNNConvolutionLayer<Dtype>::Reshape(
         this->num_output_ / this->group_, height_out, width_out,
         this->num_output_ * this->out_spatial_dim_,
         this->out_spatial_dim_, width_out, 1);
-
     cudnn::setConvolutionDesc<Dtype>(&conv_descs_[i], bottom_descs_[i],
         filter_desc_, pad_h, pad_w, stride_h, stride_w);
+    // Set cached descriptors
+    cudnn::setTensor4dDesc<Dtype>(&cached_bottom_descs_[i],
+        this->num_,
+        this->channels_ / this->group_, height, width,
+        this->channels_ * height * width,
+        height * width, width, 1);
+    cudnn::setConvolutionDesc<Dtype>(&cached_conv_descs_[i],
+        cached_bottom_descs_[i],
+        filter_desc_, pad_h, pad_w, stride_h, stride_w);
+  }
+  initialized_cached_descs_ = true;
 
-    // Have to pass full fwd/bwd cycle before taking the rest of memory
-    if (backward_passed_ctr_ > 1) {
-      // choose forward and backward algorithms + workspace(s)
-      CUDNN_CHECK(cudnnGetConvolutionForwardAlgorithm(Caffe::cudnn_handle(),
-          bottom_descs_[i], filter_desc_, conv_descs_[i], top_descs_[i],
-          CUDNN_CONVOLUTION_FWD_SPECIFY_WORKSPACE_LIMIT,
-          workspace_limit_bytes, &fwd_algo_[i]));
+  // Ask cuDNN to find the best algorithm
+  if (use_algo_seeker_) {
+    size_t workspace_limit_bytes, total_memory;
+    GPUMemory::GetInfo(&workspace_limit_bytes, &total_memory);
+    // FindEx: A workspace of size workspace_bytes is allocated for FindEx.
+    // Get: workspace_bytes is only used as a workspace limit by Get.
+    //      (no allocation happens before Get or by Get).
+    size_t workspace_bytes;
+    if (use_modest_workspace_) {
+      // In iteration 0, use a small amount of memory in order to leave
+      // most of memory for allocating layer blobs.
+      workspace_bytes = 8*1024*1024;
+    } else {
+      // Use 90% of available memory.
+      // Using all of memory may result in failure of workspace.reserve.
+      // TODO: Since 90% of memory might be too large, we can allocate
+      //       exactly how much FindEx needs by taking the maximum
+      //       workspace among all algorithms (requires an initial call
+      //       to FindEx with workspace size 0).
+      workspace_bytes = workspace_limit_bytes * 0.9;
```

Comment:
Available memory is almost always bigger (in case of unified memory, 2x bigger) than the largest contiguous block.
I suggest to implement some allocate_largest method that would catch oom and try allocating smaller blocks (requested size * 0.8 or something?) until succeeds, returning the block and the size. It should  be used further down when the memory is actually allocated.